File: frontend/js/views/bonus_grossup.js
// Bonus gross-up — the gross payment needed to net a target after tax, via
// /calc/bonus-grossup. Live as you type.

import { api } from '../api.js';
import { applyUiI18n, t } from '../i18n.js';
import { currentViewToken, viewIsCurrent } from '../app.js';
import { showToast } from '../toast.js';
import { debounce } from '../util.js';
import * as enh from '../calc_enhance.js';

const money = (n) => '$' + Number(n).toLocaleString(undefined, { minimumFractionDigits: 2, maximumFractionDigits: 2 });
const pct = (n) => Number(n).toLocaleString(undefined, { maximumFractionDigits: 2 }) + '%';
const VIEW = 'bonus-grossup';
let lastReport = null;
let lastBody = null;

export async function renderBonusGrossup(mount, _appState) {
    const tok = currentViewToken();
    if (!viewIsCurrent(tok)) return;
    mount.innerHTML = `
        <h1 class="view-title"><span data-i18n="view.grossup.h1.title">// BONUS GROSS-UP</span></h1>
        <p class="muted small" data-i18n="view.grossup.hint.intro">
            The gross payment needed so the recipient nets a target after tax — for grossing up a
            bonus, relocation, or make-whole payment. The federal supplemental-wage flat rate is 22%
            (37% above $1M); state and the 7.65% FICA stack on top. Updates as you type.
        </p>
        <div class="lpv-split">
        <div class="chart-panel">
            <h2 data-i18n="view.grossup.h2.inputs">The payment</h2>
            <form id="grossup-form" class="inline-form">
                <label><span data-i18n="view.grossup.label.net">Desired net ($)</span>
                    <input type="number" step="0.01" min="0" name="desired_net_usd" value="1000" required></label>
                <label><span data-i18n="view.grossup.label.federal">Federal rate (%)</span>
                    <input type="number" step="0.1" min="0" name="federal_rate_pct" value="22" required></label>
                <label><span data-i18n="view.grossup.label.state">State rate (%)</span>
                    <input type="number" step="0.1" min="0" name="state_rate_pct" value="5"></label>
                <label><span data-i18n="view.grossup.label.fica">Include FICA (7.65%)</span>
                    <input type="checkbox" name="include_fica" checked></label>
            </form>
            <div id="grossup-tools" class="ce-toolbar"></div>
            <button type="button" id="grossup-sens-btn" class="ce-tool" data-i18n="calc.enh.sens.run">▦ Sensitivity</button>
            <div id="grossup-sens" class="ce-sens"></div>
        </div>
        <div id="grossup-result" class="lpv-preview"></div>
        </div>
    `;
    applyUiI18n(mount);

    const form = mount.querySelector('#grossup-form');
    const hashIn = enh.readHashInputs();
    enh.prefillForm(form, hashIn);
    if ('include_fica' in hashIn) form.querySelector('[name=include_fica]').checked = hashIn.include_fica === 'true';
    const readBody = () => {
        const fd = new FormData(form);
        return {
            desired_net_usd: Number(fd.get('desired_net_usd')) || 0,
            federal_rate_pct: Number(fd.get('federal_rate_pct')) || 0,
            state_rate_pct: Number(fd.get('state_rate_pct')) || 0,
            include_fica: form.querySelector('[name=include_fica]').checked,
            fica_rate_pct: 7.65,
        };
    };
    const generate = async () => {
        const body = readBody();
        try {
            const r = await api.calcBonusGrossup(body);
            if (!viewIsCurrent(tok)) return;
            lastReport = r; lastBody = body;
            renderResult(mount, r, body, tok);
        } catch (err) {
            showToast(err.message || t('view.grossup.toast.error'), { level: 'error' });
        }
    };
    enh.mountToolbar(mount.querySelector('#grossup-tools'), { viewId: VIEW, getInputs: () => lastBody || readBody(), getRows: () => reportRows(lastReport), filename: 'bonus-grossup.csv' });
    mount.querySelector('#grossup-sens-btn').addEventListener('click', () => runSens(mount, readBody(), tok));
    form.addEventListener('input', debounce(generate, 250));
    form.addEventListener('submit', (e) => { e.preventDefault(); generate(); });
    generate();
}

function reportRows(r) {
    if (!r) return [];
    return [
        ['metric', 'value'],
        ['gross_usd', r.gross_usd],
        ['total_tax_usd', r.total_tax_usd],
        ['combined_rate_pct', r.combined_rate_pct],
        ['federal_withholding_usd', r.federal_withholding_usd],
        ['state_withholding_usd', r.state_withholding_usd],
        ['fica_withholding_usd', r.fica_withholding_usd],
    ];
}

async function renderResult(mount, r, body, tok) {
    const el = mount.querySelector('#grossup-result');
    // Line chart: required gross as the desired net sweeps 0 -> 2x (linear in net).
    const base = body.desired_net_usd || 1000;
    const xs = enh.linspace(0, base * 2, 13);
    const pts = await Promise.all(xs.map(async (n) => {
        const rr = await api.calcBonusGrossup({ ...body, desired_net_usd: n });
        return { x: n, y: rr ? rr.gross_usd : NaN };
    }));
    if (!viewIsCurrent(tok)) return;
    const chart = enh.svgLineChart(pts, { xlabel: 'net $', ylabel: 'gross $' });
    el.innerHTML = `
        <div class="chart-panel">
            <h2 data-i18n="view.grossup.h2.result">The gross-up</h2>
            <div class="cards">
                <div class="card pos"><div class="label" data-i18n="view.grossup.card.gross">Gross payment</div>
                    <div class="value pos">${money(r.gross_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.grossup.card.tax">Total tax</div>
                    <div class="value">${money(r.total_tax_usd)}</div></div>
                <div class="card"><div class="label" data-i18n="view.grossup.card.rate">Combined rate</div>
                    <div class="value">${pct(r.combined_rate_pct)}</div></div>
            </div>
            ${chart}
            <table class="data-table">
                <tbody>
                    <tr><td data-i18n="view.grossup.row.federal">Federal withholding</td><td>${money(r.federal_withholding_usd)}</td></tr>
                    <tr><td data-i18n="view.grossup.row.state">State withholding</td><td>${money(r.state_withholding_usd)}</td></tr>
                    <tr><td data-i18n="view.grossup.row.fica">FICA withholding</td><td>${money(r.fica_withholding_usd)}</td></tr>
                    <tr><td data-i18n="view.grossup.row.tax">Total tax</td><td>${money(r.total_tax_usd)}</td></tr>
                    <tr class="emph"><td data-i18n="view.grossup.row.gross">Gross payment</td><td>${money(r.gross_usd)}</td></tr>
                </tbody>
            </table>
        </div>
    `;
    applyUiI18n(el);
}

async function runSens(mount, base, tok) {
    const panel = mount.querySelector('#grossup-sens');
    panel.innerHTML = `<p class="muted small" data-i18n="calc.enh.sens.running">Computing…</p>`; applyUiI18n(panel);
    // x: federal rate 10% -> 40%; y: state rate 0% -> 13%. Output: required gross.
    const xVals = enh.linspace(10, 40, 5);
    const yVals = enh.linspace(0, 13, 5);
    const { cells } = await enh.runSensitivity({ base, xKey: 'federal_rate_pct', yKey: 'state_rate_pct', xVals, yVals, compute: (b) => api.calcBonusGrossup(b), pick: (r) => (r ? r.gross_usd : null) });
    if (!viewIsCurrent(tok)) return;
    // Lower required gross is better -> negate for shading (green = cheaper).
    panel.innerHTML = enh.renderSensitivityTable({ xVals, yVals, cells: cells.map((row) => row.map((v) => (v == null ? null : -v))), fmt: (v) => (v == null ? '—' : '$' + Math.round(-v)), xfmt: (v) => v.toFixed(0) + '%', yfmt: (v) => v.toFixed(0) + '%', xName: t('view.grossup.label.federal') || 'Fed', yName: t('view.grossup.label.state') || 'State' });
}

File: crates/traderview-core/src/bonus_grossup.rs
//! Bonus gross-up — the gross payment needed so the recipient nets a target
//! amount after tax. Used for grossing up bonuses, relocation, and other
//! make-whole payments.
//!
//! ```text
//! gross = net / (1 − combined withholding rate)
//! ```
//!
//! The federal supplemental-wage flat rate is 22% (37% above $1M); state and
//! the 7.65% FICA stack on top. The combined rate must be below 100% — there's
//! no gross-up if taxes would consume the entire payment.

use serde::{Deserialize, Serialize};

fn d_federal() -> f64 {
    22.0
}
fn d_fica() -> f64 {
    7.65
}

#[derive(Debug, Clone, Deserialize)]
pub struct GrossUpInput {
    /// Net amount the recipient should keep.
    pub desired_net_usd: f64,
    /// Federal withholding rate (22% IRS supplemental flat rate by default).
    #[serde(default = "d_federal")]
    pub federal_rate_pct: f64,
    #[serde(default)]
    pub state_rate_pct: f64,
    /// Include the 7.65% FICA (Social Security + Medicare) in the gross-up.
    #[serde(default)]
    pub include_fica: bool,
    #[serde(default = "d_fica")]
    pub fica_rate_pct: f64,
}

#[derive(Debug, Clone, Serialize, PartialEq)]
pub struct GrossUpResult {
    /// Gross payment that nets the target; 0 if the combined rate ≥ 100%.
    pub gross_usd: f64,
    /// Combined withholding rate applied, percent.
    pub combined_rate_pct: f64,
    /// gross − net.
    pub total_tax_usd: f64,
    pub federal_withholding_usd: f64,
    pub state_withholding_usd: f64,
    pub fica_withholding_usd: f64,
}

pub fn analyze(input: &GrossUpInput) -> GrossUpResult {
    let fica = if input.include_fica {
        input.fica_rate_pct
    } else {
        0.0
    };
    let combined = input.federal_rate_pct + input.state_rate_pct + fica;

    let gross = if combined < 100.0 {
        input.desired_net_usd / (1.0 - combined / 100.0)
    } else {
        0.0
    };

    GrossUpResult {
        gross_usd: gross,
        combined_rate_pct: combined,
        total_tax_usd: gross - input.desired_net_usd,
        federal_withholding_usd: gross * input.federal_rate_pct / 100.0,
        state_withholding_usd: gross * input.state_rate_pct / 100.0,
        fica_withholding_usd: gross * fica / 100.0,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn close(a: f64, b: f64) -> bool {
        (a - b).abs() < 1e-4
    }

    fn run(net: f64, fed: f64, state: f64, fica: bool) -> GrossUpResult {
        analyze(&GrossUpInput {
            desired_net_usd: net,
            federal_rate_pct: fed,
            state_rate_pct: state,
            include_fica: fica,
            fica_rate_pct: 7.65,
        })
    }

    #[test]
    fn gross_with_fica() {
        // 1000 / (1 − 0.3465) = 1530.221882.
        let r = run(1000.0, 22.0, 5.0, true);
        assert!(close(r.gross_usd, 1530.221882));
        assert!(close(r.combined_rate_pct, 34.65));
    }

    #[test]
    fn gross_without_fica() {
        let r = run(1000.0, 22.0, 5.0, false);
        assert!(close(r.gross_usd, 1369.863014));
        assert!(close(r.combined_rate_pct, 27.0));
    }

    #[test]
    fn total_tax_is_gross_minus_net() {
        let r = run(1000.0, 22.0, 5.0, true);
        assert!(close(r.total_tax_usd, r.gross_usd - 1000.0));
    }

    #[test]
    fn withholdings_sum_to_total_tax() {
        let r = run(1000.0, 22.0, 5.0, true);
        let sum = r.federal_withholding_usd + r.state_withholding_usd + r.fica_withholding_usd;
        assert!(close(sum, r.total_tax_usd));
    }

    #[test]
    fn higher_state_raises_gross() {
        let low = run(1000.0, 22.0, 0.0, true);
        let high = run(1000.0, 22.0, 10.0, true);
        assert!(high.gross_usd > low.gross_usd);
    }

    #[test]
    fn fica_toggle_changes_gross() {
        let with = run(1000.0, 22.0, 5.0, true);
        let without = run(1000.0, 22.0, 5.0, false);
        assert!(with.gross_usd > without.gross_usd);
        assert!(close(without.fica_withholding_usd, 0.0));
    }

    #[test]
    fn nets_back_to_target() {
        // Gross minus all withholding equals the desired net.
        let r = run(1000.0, 22.0, 5.0, true);
        let net = r.gross_usd
            - r.federal_withholding_usd
            - r.state_withholding_usd
            - r.fica_withholding_usd;
        assert!(close(net, 1000.0));
    }

    #[test]
    fn combined_at_or_above_100_has_no_grossup() {
        let r = run(1000.0, 100.0, 5.0, true);
        assert!(close(r.gross_usd, 0.0));
    }
}
